Add a STYLE.md file and link it from CONTRIBUTING.md

bob-beck opened the pull request:
So this has sat for a while, after clang-format paused it. 

I have now repatriated STYLE and DOCUMENTATION from the web page location to be 
in the code base where they can be up to date and easily read by a new contributor.

They are linked in from CONTRIBUTING.md and from each other appropriately. 

I have rototilled STYLE based on it's earlier content, feedback I've gotten in the
interim and things I've seen /caught up on.  I believe this makes an attempt to address
a lot of the gaps and make this a more cohesive document that is more useful to
a new contributor than the existing situation, and more reflective of both what we do
today, and where we want to get to (as a lot of this is for new code)

Substantive difference include broader guidance on macros, integer types, 
functions, and return values. 

As it appeared to still be a preference for those expressing one, This has chosen to stick
with Doxygen for internal function documentation and function headers. While I haven't
been personally doing it yet, I believe this still adds value if we start doing it, and/or fix
what's there.  

Fixes: openssl/private#851

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

bob-beck committed:
Add a STYLE.md file and link it from CONTRIBUTING.md

This is effectively, the current coding style policy web page
changed to accomodate clang-format.  This is more or less the
same file I have had in the various clang-format sample PR's since
September.

It does include an additional sections on Integers and on
Return Values that are not in the original coding style
policy

It is changed from the September version in that it does not
have mention of keeping include files self contained. I believe
that is achievable and desirable, but I think should be done
as a separate change from this.

There were a number of issues brought up in discussion of this
file in the clang-format PR's. I recorded those in 818, 819,
820, 821, 822, 823, 824, 825, and 826, which we can
link in there to the appropriate section of the document.

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,531 @@
+multiple data declarations). This leaves you room for a small comment on each
+item, explaining its use.
+
+In an effort to better translate our source code into documentation that is
```
https://github.com/openssl/private/issues/818

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,531 @@
+guarantee across different micro-architecture. Nor is it portable enough to
+meet our multi-platform support goals.
+
+##  Asserts
```
openssl/private#821

bob-beck commented on STYLE.md:
````diff
@@ -0,0 +1,531 @@
+}
+```
+
+## Typedefs
````
openssl/private#826

bob-beck commented on CONTRIBUTING.md:
```diff
@@ -75,17 +75,17 @@ guidelines:
+    There is a configuration for [clang-format](.clang-format) that will ensure
+    code is compliant with the formatting portions of the coding style guide. You should
+    configure modern tooling to make use of it and consider checking your PR's with
+    clang-format before submitting them. Where `gcc` or `clang` is available, you should use the
```
with the caveat that this document should not be a how-to guide to install such things, any more than it should be a how-to guide to use vim, emacs, git, vscode, etc.etc.

bob-beck replied:
So exactly how to do this is now in Contributing.md, and referenced from style.

levitte commented on STYLE.md:
```diff
@@ -0,0 +1,531 @@
+includes, but is not limited to, verifying that:
+
+    * non-optional pointer arguments are not NULL and
+    * numeric arguments are within expected ranges.
```
Markdown understand indentation of 4 or more spaces as a code block.  I assume that's not what you wanted here.

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,531 @@
+
+    * In the libraries when the global state of the software is corrupted and
+      there is no way to recover it
+    * In applications, test programs and fuzzers
```
I think I got em all

levitte commented on STYLE.md:
```diff
@@ -0,0 +1,689 @@
+meet our multi-platform support goals.
+
+##  Asserts
+We have 3 kind of asserts. The behaviour depends on being a debug or release
```
Please insert empty line after heading line

t-j-h reviewed (changes requested):
Just some minor tweaks / suggestions. 
I think this is pretty much ready to go in and overdue.

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,689 @@
+
+When doing arithmetic, account for overflow conditions.
+
+Except in platform specific code, do not use `ssize_t`. MSVC lacks it,
```
ssize_t is kind of evil, and tends to just introduce signed/unsigned problems.  I know why it ended up there, 
posix folks were lazy and wanted to keep things returning a negative value to indicate failure, while size_t was correctly, unsigned. 

IMO if we are pushing folks toward a cohesive "do not return magic values to indicate errors" ssize_t should not be needed in general, unless you have to use something that returns it.  We probably shouldn't be adding new stuff that needs it.

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,689 @@
+compilation is dependant on more than one variable instead of nesting
+multiple blocks.
+
+All #endif blocks must have a comment matching their #if.
```
IMO no, it ends up confusing the issue visually. You can simply bounce to the #endif and you know what the else is intending to be for.

bob-beck committed:
Modernize and update STYLE.MD, add DOCUMENTATION.MD

So this repatriates DOCUMENTATION.MD from the web page to the
code base, and links both STYLE.MD and DOCUMENTATION.MD from
CONTRIBUTING.MD

It does a large rototilling of STYLE.MD to address many of the
outstanding concerns noted when I started this before clang-format
last year, and brings us roughly in line with the things that are
addressed in similar style guides for other projects.

Most of the changed or updated reccomendations reflect what we
currently have been doing, or have expressed as a desire to
move to in the future.

Most larger "OpenSSL-isms" I've tried to explicitly call out
to make this a more cohesive and useful guide for a new contributor

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,689 @@
+course, you often don’t find out until the code is ported to other
+platforms.
+
+A final word on struct’s. OpenSSL has historically made all struct
```
fixed

bob-beck commented on STYLE.md:
````diff
@@ -0,0 +1,689 @@
+
+```
+    /**
+     * \brief ossl_snuffle_thingamabob snuffles a thingamabob from bytes of input.
````
fixed

bob-beck commented on STYLE.md:
```diff
@@ -0,0 +1,689 @@
+OpenSSL follows the
+[Webkit coding style for C code](https://webkit.org/code-style-guidelines/)
+The rest of this document describes differences and clarifications on
+top of the base guide.
```
I believe this has now basically been rolled into one, (also in CONTRIBUTING.MD) check the new wording.

levitte commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,185 @@
+-------------------------------
+
+All new public symbols must be documented in a POD manual page in the
+`doc/man3` directory. This includes both macros and functions.
```
We already include types in some newer man3 documentation, and I think we should continue to do so.

```suggestion
`doc/man3` directory. This includes types, macros and functions.
```

bob-beck replied:
aha, yes, good catch. changing

mbroz commented on CONTRIBUTING.md:
```diff
@@ -97,8 +97,9 @@ guidelines:
     (usually by rebasing) before it will be acceptable.
 
- 4. Code provided should follow our [coding style] and [documentation policy]
+ 4. Code provided should follow our [coding style](STYLE.md) and [documentation policy]
```
Add (DOCUMENTATION.md) ?

bob-beck replied:
man, how did i screw up and drop that link.  fixing

mbroz commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,185 @@
+-----------------------------------
+
+All new or modified arguments to the commands must be documented in the
+`doc/man1` directory. This documentation is in POD format.
```
Why two sentences, "All new commands or modified arguments to ..." ?

bob-beck replied:
a new command is different from an argument to a command (i.e. openssl foobar -blah -woof"  as opposed to openssl foobar, openssl barfoo,  the latter are separate comands foorbar and barfoo, whereas -blah and -woof are arguements to commands. 

Nevertheless really both need to be documented, so this wording as it exists is a bit derpy, let me fix it

mbroz commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,185 @@
+Where additional user-facing information is required, it should be
+included in the `doc/man7` section. This includes, but is not limited to:
+
+- descriptions of data structures and their usage
```
types and data structures

levitte replied:
I'd argue against types.  Types are library/API, should go in man3.

Note that data structures is a bit broader than type structures (i.e. not just `struct ...`).  PEM is a data structure.  DER is a data structure. ...

bob-beck replied:
Agree with richard. types and *public* data strutures should be go in man3.  but I'd think of a public data structure as somehting we expose publicly (i.e. pretend X509 was still non-opaque).  Yes techincally things like pem and der can also be considered data structures, but I'd sort of lean in terminology for calling those a format or serialization just to avoid confusion.    The intent here is that our public types should be documented  

So I think data strucures here for man7 is incorrect, rewording

bob-beck replied:
I think the more generic descriptions here now are more appriate for man7

mbroz commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,185 @@
+These should all be documented at the point of implementation. Follow
+the same rules and exceptions as for internal functions and structures
+above. In some cases slightly more leniency with respect to *trivial*
+can be tolerated.
```
Is the last sentence needed? There is *should* already.

bob-beck replied:
Changed.

mbroz commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,185 @@
+Very significant features and changes should be documented in the
+[NEWS.md](NEWS.md) file.
+
+In both cases, the added note should be short and to the point.
```
Maybe add that it should be written for users of library (not developers), focusing on impact, not on implementation details.

bob-beck replied:
Agree. will add.

mbroz commented on STYLE.md:
```diff
@@ -0,0 +1,1099 @@
+commit makes the diff hard to review and hard to revert.
+
+The language is C99 (ISO/IEC 9899:1999). More modern C versions
+are not yet supported on every platform OpenSSL targets and
```
Features from more modern C ...? (It can be only some specific thing, not the whole Cxx standard.)

levitte replied:
We have previously supported some more modern features, as long as they're easy to support with macros, or are fairly minimal code surrounded by appropriate guards.  But the baseline is still C99

bob-beck replied:
Given this is targeted to both us and external contributors, I do not want to be wishy washy here. If you give people an inch they will take a mile. "Oh C++23 isn't that different from C99". 

Right now we have enough difficulty getting fully on board with C99 without backpedalling :) 

If at a later date we find we're doing well at C99 and want to introduce some more things we can make that
decision and change this.  Today the resolution of my c99 issue was tomas agreeing that C99 could be used on master and be done with it, we deal with merging to older stuff.

levitte replied:
Sounds like a plan.  👍🏼

mbroz commented on STYLE.md:
```diff
@@ -0,0 +1,1099 @@
+as `long` / `long long`: their widths vary by platform.
+
+Bit shifts should be performed on unsigned operands.
+Left-shifting a signed value is undefined behaviour when the
```
Maybe add specific section describing than no undefined behaviours should be introduced (and maybe how to handle known issues)?

bob-beck replied:
At that point we might end up repeating the entire C standard, which I don't think we want to do here. 

I've tried to limit the specific call outs like this to be similar to what *other* major C projects also explicitly call
out, and what I have seen be a problem in our own codebase myself.  I don't think STYLE should become a reference for all possible C ub behaviour, and I think these callouts make more sense to the reader in a context ("oh remember this when doing integer types" - "oh remember this about macro use") than an entire section of boring stuff that is unrelated UB behaviours all lumped together that I must find everything and understand it without context.

mbroz commented on STYLE.md:
```diff
@@ -0,0 +1,1099 @@
+
+int f(void *p)
+{
+    RETURN_IF_NULL(p);   /* may return from f() -- not visible at the call site */
```
Mark it as /* This is bad . */ as above.

bob-beck replied:
I think in the context of the paragraph above that is driving this example, and the comment line
saying what is happening, I don't think adding another /* this is bad */ line really contributes anything here.

mbroz commented on STYLE.md:
```diff
@@ -0,0 +1,1099 @@
+### Centralised exiting of functions
+
+When a function exits from multiple locations and some common
+work (such as cleanup) has to be done at every exit, use `goto`
```
Maybe add another section discussing where goto should not be used?
(like jumping bakwards, out of some scope etc)

bob-beck replied:
I think that starts to feel like adding more and more prohibitions to make it feel like a C textbook or a programming class..  I might go the other way on this one but I sort of kept to the "here is how to do the
thing" expclitly calling out the goto usage, which most people will avoid by default, rather than wasting the
readers time repeating the same old hoary don't use goto bullshit we've all seen a thousand times over.

If we really think repeating that makes the document better, I could be convinced to change my mind, but
right now I don't think it does.

mbroz commented on STYLE.md:
```diff
@@ -0,0 +1,1099 @@
+-----------------------
+
+The only reason for processor-specific code in OpenSSL is
+performance. Every processor-specific path must have a
```
What about constant-time execution?

bob-beck replied:
I disagree, you can also do that in C. read below for the requirement that "you must make your cryptographic operations in assembly constant time" though.

andrewkdinh commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,185 @@
+and will be brought up to standard gradually.
+
+Any non-trivial change to existing code must bring the affected code into
+conformance with this policy as part of the same change. In particular,
```
Should we mention that it's ok if it's a separate PR, just as long as it's done?

bob-beck replied:
I think I'm trying to keep the "how" out of this, and mostly the "what" in this.  Talk of bringing things into conformance with a separate PR is in STYLE.md

bob-beck committed:
Change the Doxygen doc requirement for non public functions

To be at the prototype site in the internal header file.

The major reason *Why* we would like to have Doxygen style comments
describing what internals do is so that they work with modern IDE's
since most common ones support them.

It's wonderful to be looking at an internal function, thinking "wtf is this",
and be able to hover over it and - boink - up comes the docs. This
typically only works (or works better) when the Doxygen comment is
at the prototype site, not if it is at the implementation site.

This also reinforces the requirement that "yes you do this for shared
functions but you don't need to for statics".

bob-beck commented on DOCUMENTATION.md:
```diff
@@ -98,16 +106,17 @@ obvious. Some representative examples:
-above. In some cases slightly more leniency with respect to *trivial*
-can be tolerated.
+These should all be documented at the point of definition. Follow the
+same rules and exceptions as for internal items above. In some cases
```
I have considered just saying "doing this is optional for static functions" - leaving it to the discretion of the implementor.   at least until such a time as we have better coverage for this for internal functions. 

But I've decide to leave the "ok you can just omit the trivial ones" with the call out here. 

What's everyone else think?

bob-beck replied:
The corrollary to this is that in a perfect world where everyone has a consistent view of trivial, this never becomes a review issue, since the nontrivial ones are documented and the trivial ones are not and everyone agrees.

Reality is we know that won't happen, so I would anticipate review time getting spent on this, which
in a perfect world might be a good thing if not delaying useful changes to excess, however, given the state we are in today, I'd contend that spending
our time on such reviews today might not be the best use of our time.   Hence my temptation for now
to just say "You never need to do this for statics" - so when people do it's a bonus, but if they don't
it's not yet another thing to go back and forth on that IMO, today, is low value.

esyr replied:
Well, what's the issue there? I don't think anyone would be arguing against documenting something that is deemed too trivial for him/her, so the only outcome is that statics will be documented to the highest requirement among reviewers.

bob-beck replied:
Actually that's exactly the problem "the highest standard among the reviewers" is effecively arbitrary, and adds latency, review effort, and cycles to contributions. 

I would like to have a clear rule here for the moment that establishes what the requirements are and what the non-requiremens are, to avoid differing standards of review, and corresponding frustration. 

I think a workable place to cut that off today would be "documenting statics in this manner is optional". My concern here is the trivial just adds back and forth review cycles for little real value today.

bob-beck replied:
Having done this for a little bit experimentally myself, I think this is the right boundary to avoid unecessary fluff and review noise for trivialities at first.  I think if we get better at this we can revisit and change our minds later.

bob-beck commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,194 @@
+items declared in any header that is not part of the public API.
+These include items declared in:
+
+- `include/internal/` (shared across subsystems);
```
might end up looking something like : https://github.com/openssl/openssl/pull/31488

nhorman commented on CONTRIBUTING.md:
```diff
@@ -97,8 +97,9 @@ guidelines:
- 4. Code provided should follow our [coding style] and [documentation policy]
-    and compile without warnings.
+ 4. Code provided should follow our [coding style](STYLE.md) and
+    [documentation policy](DOCUMENTATION.md) and compile without warnings.
```
is it worth mentioning that it should compile without warnings under a --strict-warnings config here?

bob-beck replied:
yeah it hints but we should be explicit

nhorman commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,195 @@
+is a navigation aid pointing to the canonical reference documentation
+in the corresponding POD file; see [STYLE.md](STYLE.md) for the
+doxygen form.
+
```
I presume that a doxygen block with a single @see reference here is sufficient (so as to avoid duplication usage descriptions that might drift apart)?

bob-beck replied:
yes. 100% hence the discussion of "do not replace the pod file"

nhorman reviewed (dismissed):
just the two nits, I don't think its worth holding this up for it though.

mbroz reviewed (dismissed):
Well, there will be perhaps always some details where people disagree.
But let's start with something in tree, it is really waiting too long in the queue.

openssl-machine commented:
This pull request is ready to merge

andrewkdinh reviewed (approved):
We should probably should just get this in as @mbroz notes. We also need to remove the current coding style guide at https://www.openssl-library.org/policies/technical/coding-style/

mattcaswell commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+just to convert to `int` at the API boundary adds friction
+without enough benefit.
+
+Except in platform-specific code, do not use `ssize_t`; MSVC lacks
```
We have `ossl_ssize_t` for cross platform usage

bob-beck replied:
I still don't think that changes my mind here.  while I think it's fine to have that if we need it in places to deal with other things returning ssize_t things, I don't think we should be doing that in new code.

mattcaswell commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+Prefer functions over function-like macros. Do not optimise for
+function-call overhead without first measuring with a function
+implementation; if the function is hot enough to need inlining,
+mark it `inline` rather than converting it to a macro.
```
`ossl_inline`

bob-beck replied:
Actually now that we are agreeing to be fully C99 (see above) we should be getting rid of this. 

"inline" is fully supported in C99, so there's no reason to have a custom one of these anymore

bob-beck replied:
... and having looked that's well and truly entwined in a few things right now, so I think that is in fact a bridge too far, let's just stick to ossl_inline until we're ready.. changing this

mattcaswell commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+else /* This now breaks. */
+
+/* This is also bad, because now you have to omit the semicolon. */
+#define KERMIT(x) { muppet((x)); frog((x)); green((x)) }
```
```suggestion
#define KERMIT(x) { muppet((x)); frog((x)); green((x)); }
```

bob-beck replied:
I've mostly left these out of these example to match the initial define above. 

Strictly speaking the semicolon is not needed on the last statement of a block (although we almost alway put one there.   I had left these without it to most closely match the progression from the first example..

Mostly I've continued down the garden path from the first example, where the semicolon is omitted
but it still doesn't work because it isn't a block, so I think I'm more happy (in this example) of sticking to the "no semicolon at the end" because for single statement macros we want the same.

bob-beck replied:
I am mildly on the fence about it though, because, were I to write that macro, I would put the semicolon on the last statement of the block.

nhorman replied:
why not write it as a do { ... } while(0) loop?  That enforces the need for a semicolon at the instantiation site

bob-beck replied:
See below.  The point of the example is to show the wrong ways of doing it.

bob-beck replied:
it is done as a while below this.

mattcaswell replied:
> Strictly speaking the semicolon is not needed on the last statement of a block

Errr....yeah it is?? Unless of course `green` itself is a macro that does emit the `;` - but otherwise this is not valid C.

mattcaswell commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+else
+
+/* This works. */
+#define KERMIT(x) do { muppet((x)); frog((x)); green((x)) } while (0)
```
```suggestion
#define KERMIT(x) do { muppet((x)); frog((x)); green((x)); } while (0)
```

bob-beck replied:
similarly.

mattcaswell commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+
+### Do not include files as multi-line macros
+
+Do not put code in a file and include it inline:
```
We do this in some cases where the included code is generated and we don't want to pollute the hand written code with generated stuff.

bob-beck replied:
I agree there are exceptions to this rule (although often in this case, you can simply put the generated code into a file and call it without using #include) and I think in the rare situations we want to do this we can figure out whether to do it this way or not.

bob-beck commented:
> We should probably should just get this in as @mbroz notes. We also need to remove the current coding style guide at https://www.openssl-library.org/policies/technical/coding-style/

there's also the documentation web page, etc.  I'm not changing that here, it has to happen somewhere else.

esyr commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,195 @@
+The project's documentation is about making the libraries and tools more
+accessible to our users and making the code more maintainable. This policy
+applies to new submissions; existing code does not uniformly conform to it
+and will be brought up to standard gradually.
```
`s/standard/the standard/`

bob-beck replied:
I disagree, the "the" is unnecessary.

esyr commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,195 @@
+The allowed exceptions are:
+
+- guard macros preventing a header file being included twice
+- new `OPENSSL_NO_DEPRECATED_` macros
```
That is not correct, they are documented in `doc/internal/man7/deprecation.pod`.

bob-beck replied:
ugh yes, fixing

esyr commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,195 @@
+
+Each public function's declaration in its public header must carry a
+doxygen comment block. The block's `@see` must include the function's
+own manual page (`name(3)`) and may include additional manual pages
```
Not sure about "own manual page", as OpenSSL tends to document groups of functions on a single page, maybe just "the manual page in which the function documented"?

t8m replied:
Also I can guarantee that this won't be followed fully at least from the beginning. All such major changes to style or documentation guide would have to be explicitly communicated to all the committers so they do not forget about it during reviews.

There should be also some kind of doc-nits check for this.

bob-beck replied:
I disagree this won't be followed. it's pretty easy to add a check for this to CI to ensure new code has this.

esyr commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,195 @@
+[STYLE.md](STYLE.md). Comments inside the `asm()` body itself use
+plain C `/* */` comments.
+
+There are no *trivial* exceptions for assembly code.
```
`s/*trivial*/"trivial"/`

esyr commented on DOCUMENTATION.md:
```diff
@@ -0,0 +1,195 @@
+
+The language used for documentation shall be *British English*.
+
+In general the language, abbreviations, layout and formatting should also
+correspond to the
+[LDP](https://openssl-library.org/policies/general/glossary/#ldp)
+guidelines.
```
I've tried to figure out what that actually means and ended up with nothing: section 4.1.1 "Writing Style and Style Guides" of LDP Author Guide[1] just says "[t]he LDP does not require a specific style guide; however, you should use a consistent style throughout your document" and defers it to reviewer guide afterwards: "[t]he [Reviewer's HOWTO](http://tldp.org/HOWTO/LDP-Reviewer-HOWTO/) currently lists a number of conventions for LDP documents and it is as close as it gets to an official LDP Style Guide".  The latter has a "Language Review" section[2], but it also is not terribly helpful in terms of review of specifics of spelling and punctuation in order to keep it consistent corpus-wise.  I think we should settle on some specific guide ([3], maybe? Haven't actually look into it) and stick to it.

[1] https://tldp.org/LDP/LDP-Author-Guide/html/sg-writingstyle.html#writing-style
[2] https://tldp.org/HOWTO/LDP-Reviewer-HOWTO/languagereview.html
[3] https://www.thebritishacademy.ac.uk/publishing/resources/style-guide/

bob-beck replied:
Yeah that's just our internal link, which is I agree, not super useful at the moment, but it's what was worth putting in here.  

3) might be a good choice here, as you suggest.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+Applicability
+-------------
+
+New code in OpenSSL is expected to follow the conventions in this
```
"conventions" "outlined", "laid out"? Feels like a missing verb.

bob-beck replied:
conventions is a plural noun, these are them. does not feel like that to me.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+code. Combining a compliance sweep with a behaviour change in one
+commit makes the diff hard to review and hard to revert.
+
+The language is C99 (ISO/IEC 9899:1999). More modern C versions
```
Should it refer to `NOTES-C99.md`?

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+Outside spec-mirroring, single letters are appropriate only as
+loop counters (`i`, `j`, `k`).
+
+For functions, OpenSSL names follow a `PREFIX_object_action()`
```
`PREFIX_action_object()`, no?

t8m replied:
No, although in many cases the `PREFIX == OBJECT` or the OBJECT is in uppercase.

So it is actually debatable whether this shouldn't be `PREFIX_OBJECT_action()`.

bob-beck replied:
or PREFIX_[OBJECT}_action perhaps.. 

I'm not too fussy on what's decided here, it's more a guidline for new stuff, as we know many old things will not fall into this mould.  it's sort of "is there something most of the stuff does today. 

Probably you're correct that PREFIX_OBJEC_action might be closee..

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+For functions, OpenSSL names follow a `PREFIX_object_action()`
+shape: an uppercase subsystem prefix, followed by an object or
+context, followed by an action in lowercase with underscores.
+Examples: `X509_get_subject_name`, `EVP_PKEY_sign`,
```
`X509_get_subject_name` is bad in accordance with the `_get[01]` convention below.

t8m replied:
Not really as the action is `get_subject_name` and the object is the prefix.

bob-beck replied:
yeah, this is kind of a crappy example. we are, in some sense all over the map, so I guess I'll pick a better one and warn that this is to some extent aspirational and the existing naming has years of baggage behind it and does not always conform.

bob-beck replied:
applying a fix for this and the above comment. I think it reads better.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ */
+
+/**
+ * @def MAX(x, y)
```
`@def` is unnecessary here.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+#define MAX(x, y) (x > y ? x : y)
+
+/**
+ * @struct foo_st
```
I guess having `@typedef FOO` should be helpful here, but need to check how doxygen reacts if both are provided.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * @param y integer input value
+ * @returns the maximum of x and y
+ */
+#define MAX(x, y) (x > y ? x : y)
```
That's a terrible example of a macro function, please add parentheses around arguments, at least.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+/**
+ * @struct foo_st
+ * @brief Description of the foo_st struct.
+ * Optional more detailed description here.
```
There should be an empty line after `@brief`. Also, I vaguely remember that doxygen auto-briefs the first sentence of a description, but need to check.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * Document a macro that returns the maximum of two inputs.
+ * @param x integer input value
+ * @param y integer input value
+ * @returns the maximum of x and y
```
Also, an important question, `@return` or `@returns`?

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+
+/**
+ * @brief Describe the function add briefly.
+ * Add a more detailed description here, like sums two inputs and
```
Missing empty line.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * @brief Describe the function add briefly.
+ * Add a more detailed description here, like sums two inputs and
+ * returns the result.
+ * @param a input integer to add
```
I also prefer an empty line after the description, but don't recall it being required.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * - Calvin: input to be transmogrified
+ * - Hobbes: transmogrified output (caller-allocated)
+ *
+ * @see https://www.example.org/rfc/rfc31337.html#section-1.2.3
```
I think it won't hurt putting a link to `https://www.rfc-editor.org/rfc/rfc31337.html` here, so it can be copy-pasted as-is.

bob-beck replied:
It is not a real RFC, so It's example.org just so people don't go hitting the ietf site for it.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * @brief Transmogrify Calvin into Hobbes per RFC 31337 section 1.2.3.
+ *
+ * Variable naming follows the spec:
+ * - Calvin: input to be transmogrified
+ * - Hobbes: transmogrified output (caller-allocated)
```
Why not `@param` syntax?

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * @see https://www.example.org/rfc/rfc31337.html#section-1.2.3
+ */
+int transmogrify(const uint8_t *Calvin, size_t Calvin_len,
+                 uint8_t *Hobbes, size_t Hobbes_len);
```
So, does that mean that the section about variable naming does not apply to spec-mirroring parameters?

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+ * @param b input integer to add
+ * @returns the sum of a and b
+ */
+int add(int a, int b);
```
Since it has no prefix, it should be `static`.

t8m replied:
or better add a ossl_ prefix

esyr commented on STYLE.md:
````diff
@@ -0,0 +1,1093 @@
+```c
+/**
+ * @brief One-line summary of what the function does.
+ * @see X509_verify_cert(3)
````
Not exactly sure what is the value of such reference. To make visible that the function is documented?

t8m replied:
Yeah, this is newly added requirement and I disagree that will be followed.

bob-beck replied:
Again, as mentioned before, it's pretty easy to add something to ci / doc-nits to ensure this is followed for new code.

bob-beck replied:
Mostly the value is so that things that readers see immediately where to go, and doxygen integrated things also pick it up and can link you to the man page directly.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+Flexible array members (C99 trailing `[]`) are permitted and
+preferred over the older `[1]` "struct hack" for variable-length
+trailing data. Remember that `sizeof(struct)` does not include the
+flexible member; allocate the trailing data explicitly when the
+struct is created.
```
We probably need helpers for that, similarly to the ones Linux has.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+
+Avoid bitfields in structs: their layout is implementation-defined
+and varies across compilers and ABIs. Use explicit shifts and
+masks on a regular integer where bit-level positioning matters.
```
While technically it's true, in practice bit fields work rather well, so I don't agree with this statement, when bit fields are used in internal structs. Maybe `s/bitfields in structs/bitfields in public structs/`?

t8m replied:
Absolutely agree with @esyr We should not disallow bitfields in internal structs.

bob-beck replied:
Just as in the macros case, nothing in here says "prohibited" it says "avoid".

bob-beck replied:
so here, I would say, as with macros, this guidance here is to avoid them. We don't prohibit them, because like macros, there are cases where they are useful, and done right work just fine.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+for struct initialisation, particularly where they make the field
+assignments self-documenting.
+
+A trailing comma in an initializer list is a layout hint to
+`clang-format`: with it the list is kept one element per line;
+without it the formatter may pack the list onto fewer lines.
+Most of the time you do not want a trailing comma; omit it
+unless you specifically want to lock the one-per-line layout
+(for example, in a multi-row table of values).
```
This is just ugly.

bob-beck replied:
this is just documenting the reality of clang-format as it is today. 

I personally think webkit style is ugly anyway :) - but I don't care as long as it's consistent.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+Integers
+--------
+
+Prefer explicitly-sized integers over generic C ones where the
```
extraneous `the`?

bob-beck replied:
not in my english class, but yes, you could omit it, but really, this is just a nit.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+platforms. Use `int32_t`, `int64_t`, `size_t`, or another
+`<stdint.h>` type as appropriate.
+
+Sizes are `size_t`. When converting to or from `int` for legacy
```
"Sizes of in-memory objects", specifically. For storage there's `off_t` (or `uint64_t`, which is enough for everybody).

bob-beck replied:
yeah I'll be more explicit here.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+Do not retroactively narrow existing integer fields in legacy
+structs; this risks ABI breakage.
+
+When doing arithmetic, account for overflow.
```
Refer to `internal/safe_math.h` here?

bob-beck replied:
No, because it's not documented.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+typically take the form `UINT32_C(1) << n` or `(uint32_t)x << n`.
+
+In structs that are retained across the lifetime of a connection,
+new integer fields whose value range is known should use a smaller
+integer type (`uint8_t`, `uint16_t`) where doing so is
```
"naturally aligned smaller integer type".

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+parameters; those use the conventional integer types so callers
+are not forced to deal with narrow types.
+
+Do not retroactively narrow existing integer fields in legacy
+structs; this risks ABI breakage.
```
I think this rule has a lot of exceptions.

bob-beck replied:
It does, and I've contemplated pulling it out, but historically it's been a win in other places.  (Boring shaved off a lot of memory usage with this)

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+
+When doing arithmetic, account for overflow.
+
+Use `int` with `0` / `1` for boolean values, both in public API
+and internal code. Do not introduce `<stdbool.h>` for new code;
+the public API convention is `int`, and using `bool` internally
+just to convert to `int` at the API boundary adds friction
+without enough benefit.
```
I disagree here, `bool` clearly signals the `success`/`failure` semantics for return values, and works well with flags.

bob-beck replied:
I would normally agree with you here eugene, but IMO we have far too much public API that goes the other way and uses int, and I think this just becomes type conversion madness, for no real benefit.

esyr commented on STYLE.md:
````diff
@@ -0,0 +1,1093 @@
+
+```c
+#if !defined(OPENSSL_FOO_H)
+# define OPENSSL_FOO_H
````
`clang-format` will remove the space before `define`, so this is not a stylistically correct example.

bob-beck replied:
yes, some of these were brought in from changes from before clang-format ;)

bob-beck replied:
will fix

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+compilation is dependent on more than one variable, without
+nesting multiple blocks.
+
+All `#endif` blocks must have a comment matching their `#if`:
```
"All conditional blocks that span more than three lines require a matching comment for the `#endif` directive."

esyr commented on STYLE.md:
````diff
@@ -0,0 +1,1093 @@
+```c
+#if defined(OPENSSL_NO_FOO)
+static ossl_inline int foo_init(void) { return 1; }
+static ossl_inline void foo_cleanup(void) {}
````
`clang-format` will add a space between curly braces, this is not a stylistically correct code.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+#if defined(OPENSSL_NO_FOO)
+static ossl_inline int foo_init(void) { return 1; }
+static ossl_inline void foo_cleanup(void) {}
+#else
```
Personally, I prefer adding `/* !defined(OPENSSL_NO_FOO) */` here, but not sure if it is worth requiring something like that in the style guide.

esyr commented on STYLE.md:
````diff
@@ -0,0 +1,1093 @@
+example:
+
+```c
+#define BOB(blah) ((blah) + 42 - 23 / (blah))
````
Parenthesisation won't help with parameters with side effects, so that should be mentioned in macro's documentation.

esyr commented on STYLE.md:
````diff
@@ -0,0 +1,1093 @@
+
+```c
+/* This is bad. */
+#define KERMIT(x) muppet((x)); frog((x)); green((x))
````
Double parentheses around `x` are not needed here.

esyr replied:
`clang-format` will reformat this, so this is not a stylistically correct example.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+else
+
+/* This works. */
+#define KERMIT(x) do { muppet((x)); frog((x)); green((x)); } while (0)
```
`clang-format` will reformat this, so this is not a stylistically correct example.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+else /* This now breaks. */
+
+/* This is also bad, because now you have to omit the semicolon. */
+#define KERMIT(x) { muppet((x)); frog((x)); green((x)); }
```
`clang-format` will reformat this, so this is not a stylistically correct example.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+#define SQUARE(x) ((x) * (x))
+
+int n = 1;
+int result = SQUARE(n++);   /* expands to ((n++) * (n++)) -- evaluates twice */
```
This is also UB.

esyr commented on STYLE.md:
````diff
@@ -0,0 +1,1093 @@
+```c
+/**
+ * @brief ossl_snuffle_thingamabob snuffles a thingamabob from bytes of input.
+ * If a valid thingamabob is snuffled, the result is stored in
````
Missing empty line.

esyr commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+The extended function preserves the existing parameters in their
+existing order. New parameters may be inserted at any position
+(they do not have to be at the end); parameters that are no
+longer needed may be removed.
```
I would mention design considerations wrt extensible interfaces, akin to [1][2][3], but that's only me, I guess.

[1] https://www.kernel.org/doc/html/latest/process/adding-syscalls.html#designing-the-api-planning-for-extension
[2] https://lwn.net/Articles/704479/
[3] https://lpc.events/event/7/contributions/657/attachments/639/1159/extensible_syscalls.pdf

bob-beck replied:
I think that's probably going a bit far for "style" and more into design in general, which i'd say is kind of out of scope for this.  while I agree there's elements in here that affect design considerations, I don't really want to turn this into a software design textbook, and we have to draw the line somewhere.

esyr reviewed (changes requested):
Some minor correction in doxygen syntax, mostly.

openssl-machine commented:
This pull request is ready to merge

t8m commented on CONTRIBUTING.md:
````diff
@@ -165,7 +164,7 @@ guidelines:
 
     ```
-        Fixes #XXXXX
+        Fixes: LINK
````
Hmm, I wonder, who will follow this requirement.

bob-beck replied:
This is not a new requirement. So I don't understand the unhelpful speculation here. 

If you assume people will not follow the guide, and reviewers won't call it out, you should delete the style guide entirely.

levitte replied:
I already do, althought I use `Resolves:` rather than `Fixes:`.  That should be fine, github considers them equally valid (along a bunch more).

I don't see a reason for us to be particularly anal about the exact keyword, but I would press on form, as this makes for a proper git commit trailer, which `Fixes #XXXXX` is *not* (lack of colon, and just a number is utter rubbish if you don't know it's github specific)

bob-beck replied:
and since our project has multiple "issues" in different projects, this sometimes ends with confusion if you use just a number. github allows this to be the full URL to the issue, and then it unabmiguously works.

t8m commented on STYLE.md:
```diff
@@ -0,0 +1,1093 @@
+up to standard gradually; non-trivial changes to existing code
+should bring the affected area into compliance.
+
+When bringing an area into compliance as part of a larger change,
+do so in a separate commit -- typically one that lands first, so
+that the substantive change then operates on already-compliant
+code. Combining a compliance sweep with a behaviour change in one
+commit makes the diff hard to review and hard to revert.
```
Add: "Do not try to bring code into compliance in pull requests that are fixing bugs where the fix is expected to be merged into stable branches."

bob-beck replied:
I'd say in general "don't do this if you are fixing a bug, do the minimal bug fix" so yes, it's worth adding that here I  believe.

bob-beck replied:
Added specifics to address this.

t8m reviewed (commented):
I think this should not be merged as is. We will need to have broader agreement than just two reviewers approving.

bob-beck committed:
Make the examples in SYTLE and DOCUMENTATON compliant.

Specifically bring them closer to the suggested naming conventions
and make them clang-format compliant.

bob-beck commented:
> I think this should not be merged as is. We will need to have broader agreement than just two reviewers approving.

So this has been rebased with many of the useful nits addressed, some I didn't agree with. 

So I think this has been discussed a lot with many folks who have skin in the game. We can continue to discuss forever or decide to put it in as an improvement, and then if anything in practice becomes unwieldy or not useful, well, we change it. and allow the document to adapt rather than attempting to have it in a situation as it is today where a lack of change cause us to do paralysis of analysis and try to achieve perpetual perfection in one fell swoop.

openssl-machine commented:
This pull request is ready to merge

nhorman commented:
merged to master, thank you
